include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp

```diff
@@ -28,12 +28,69 @@
 #include <seqan3/range/view/to_char.hpp>
 #include <seqan3/std/ranges>
 
-namespace seqan3
+// ---------------------------------------------------------------------------------------------------------------------
+// unaligned_seq transformation trait
+// ---------------------------------------------------------------------------------------------------------------------
+
+namespace seqan3::detail
+{
+
+//!\brief Helper function to deduct the unaligned sequence type from an aligned sequence container.
+template <template <typename...> typename container_type, typename seq_alph_t, typename ...rest_t>
+//!\cond
+    requires container_concept<container_type<gapped<seq_alph_t>, rest_t...>>
+//!\endcond
+auto get_ungapped_type(container_type<gapped<seq_alph_t>, rest_t...>)
+    -> container_type<seq_alph_t, rest_t...>;
+
+//!\overload
+template <template <typename...> typename container_type,
+          template <typename...> typename allocator_type,
+          typename seq_alph_t>
+//!\cond
+    requires container_concept<container_type<gapped<seq_alph_t>, allocator_type<gapped<seq_alph_t>>>>
+//!\endcond
+auto get_ungapped_type(container_type<gapped<seq_alph_t>, allocator_type<gapped<seq_alph_t>>>)
+    -> container_type<seq_alph_t, allocator_type<seq_alph_t>>;
+
+//!\brief Default transformation trait that shall expose the unaligned sequence type of t when specialised.
+template <typename t>
+struct unaligned_seq
+{};
+
+//!\brief Exposes the unaligned sequence type given an aligned sequence container type.
+template <typename t>
+//!\cond
+    requires container_concept<t>
+//!\endcond
+struct unaligned_seq<t>
+{
+    using type = decltype(get_ungapped_type(std::declval<t>())); //!< The unaligned sequence type of t
```

Comment:
The function parameter type is specialized for the gapped alphabet type so there is no need to require it?
`constexpr auto remove_gap_from_value_type(container_type<gapped<seq_alph_t>, rest_t...>)`

Reply:
I think the compiler error will be not as easy to read, as he will complain about not finding viable overloads. 
But I am not sure. Could you test it maybe?

Reply:
<details> <summary>Error message like it is now: (click me) </summary> 

```
In file included from /home/svenja/AGABI/seqan3/test/unit/alignment/../alignment/aligned_sequence_test_template.hpp:12:0,
                 from /home/svenja/AGABI/seqan3/test/unit/alignment/aligned_sequence_test.cpp:14:
/home/svenja/AGABI/seqan3/include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp: In instantiation of ‘struct seqan3::detail::unaligned_seq<std::vector<seqan3::dna4> >’:
/home/svenja/AGABI/seqan3/test/unit/alignment/aligned_sequence_test.cpp:43:58:   required from here
/home/svenja/AGABI/seqan3/include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp:71:53: error: no matching function for call to ‘remove_gap_from_value_type(std::vector<seqan3::dna4>)’
     using type = decltype(remove_gap_from_value_type(std::declval<t>()));
                           ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
/home/svenja/AGABI/seqan3/include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp:43:16: note: candidate: template<template<class ...> class container_type, class seq_alph_t, class ... rest_t>  requires  container_concept<container_type<seqan3::union_composition<seq_alph_t, seqan3::gap>, rest_t ...> > constexpr container_type<seq_alph_t, rest_t ...> seqan3::detail::remove_gap_from_value_type(container_type<seqan3::union_composition<seq_alph_t, seqan3::gap>, rest_t ...>)
 constexpr auto remove_gap_from_value_type(container_type<gapped<seq_alph_t>, rest_t...>)
                ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/svenja/AGABI/seqan3/include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp:43:16: note:   template argument deduction/substitution failed:
/home/svenja/AGABI/seqan3/include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp:71:53: note:   mismatched types ‘seqan3::union_composition<seq_alph_t, seqan3::gap>’ and ‘seqan3::dna4’
     using type = decltype(remove_gap_from_value_type(std::declval<t>()));
```

</details>

<details> <summary>Error message with an additional requires "is_gapped_alphabet": (click me) </summary> 

```
/home/svenja/AGABI/seqan3/test/unit/alignment/aligned_sequence_test.cpp: In member function ‘virtual void hi_hi_Test::TestBody()’:
/home/svenja/AGABI/seqan3/test/unit/alignment/aligned_sequence_test.cpp:43:60: error: ‘type’ in ‘struct seqan3::detail::unaligned_seq<std::vector<seqan3::dna4> >’ does not name a type
     using type = detail::unaligned_seq<std::vector<dna4>>::type;
```

I am not sure but the first is maybe even more descriptive as it tells you that dna4 is no gapped alphabet while the second just fails because the chosen struct (chosen by overload resolution when the requires fails) is the default one without a type member.

@rrahn what do you think?

Reply:
Mhmm, I thought that the requires clause will emit a constraint not satisfied compiler error pointing to the is_gapped_alphabet expression. Does this only works with concepts?